### DISCUSSION TITLE
Change svg image colors at runtime?

### DISCUSSION THREAD

**Discussion by ms-somanna** [ElinamLLC/SharpVectors#163] [Ideas]
Alright, so I have a bunch of svg images that I've added to my application as resources. I'm using these images as a icons for several buttons by adding these images to svg viewport within the buttons. These images are multicolored and a part of it contains white color. Now what I want to achieve is during my application's runtime when I change the application theme (which I can handle) I want the white parts of the svg images in the buttons to become black. Is this possible with sharp vectors? If yes, how?

**Comment by paulushub** [comment]
@SwaroopSomanna I have looked into ways to support themes, but gave up! There is no standard way to handle it, so instead I provided several means to implement it.
For your case, is it just all white colors must become black or these are defined with style Id and/or class?

**Comment by ms-somanna** [comment]
> For your case, is it just all white colors must become black or these are defined with style Id and/or class?  

No style ID or classes is used. I just want  all white colors to become black. That's all.

**Comment by paulushub** [comment]
@SwaroopSomanna Sorry for the late response, please can you provide a sample SVG file for illustration?

**Comment by ms-somanna** [comment]
> @SwaroopSomanna Sorry for the late response

That's ok

**Comment by ms-somanna** [comment]
> @SwaroopSomanna please can you provide a sample SVG file for illustration?  

Now I'm using monochrome icons with blue foreground that works best for both black and white themes. Thanks anyway.

**Comment by GF-Huang** [comment]
So how to change foreground at runtime???

**Reply by paulushub** [reply]
@GF-Huang What is your use case? Currently, there is no direct way provided by the controls or markup extensions, except through user-css, supported through settings object.

If you are using the converters, you can user-css, element visitors etc. I wish to provide color remap, but I need practical use cases. Unfortunately @SwaroopSomanna switched to a different approach and we could not continue to efforts.

The user-css is only applicable if you have complete control over the design of the SVG file, defining CSS classes and IDs that can be restyled. We could add inherited dependency properties to make this easier.

For element visitors, you will need access to the document object before rendering, and we have started adding events to make this easier. Could be possible through IoC as well.

Color/Brush remap will be easier to work with and be supported through dependency injection, we have already added a tiny IoC library (part of the model library) to make this possible. But this, I think, are only use for monochrome or duotone images (like icon backgrounds and/or foregrounds).

Something more WPF-stylish could also be useful, since this could flow with the theming of the whole application.

**Reply by GF-Huang** [reply]
```xaml
...
<TabControl>
    <TabItem Header="{svg:SVGImage Source=xxx, Foreground={DynamicResource xxxColorOrBrush}}">
    </TabItem>
</TabControl>
```

**Reply by GF-Huang** [reply]
I think your design has gone too far for most users, who just want to download some `*.svg` files from the Internet and simply display them in the UI.

**Reply by paulushub** [reply]
The source is SVG, and there is no "foreground" accept as part of a style and fill or stroke.

**Reply by GF-Huang** [reply]
Just add a `foreground` property binding to the `Path.Fill`.

**Reply by GF-Huang** [reply]
Take a look of [MahApps.Metro.IconPacks](https://github.com/MahApps/MahApps.Metro.IconPacks), it's very easy to use.

They just convert the svg to xaml path.

![image](https://user-images.githubusercontent.com/4510984/110964432-25efcf80-838e-11eb-8c79-7f3441e8f4cd.png)

![image](https://user-images.githubusercontent.com/4510984/110964624-559ed780-838e-11eb-8871-327fdd22dfde.png)

**Reply by paulushub** [reply]
I understand most are simply displaying svg icons, but this library was designed for complete support of SVG rendering with scripting.
I was thinking an alternative like [SVGImage](https://github.com/dotnetprojects/SVGImage) could fill that gap, but it lacks support for adequate CSS and might not satisfy most needs. I am currently thinking about SharpVectors.Lite that will reside in a single DLL for the WPF users. However, maintaining multiple libraries is a problem I need to resolve! 

> Just add a foreground property binding to the Path.Fill.

That could be simply to do, will look into it.

**Reply by paulushub** [reply]
> Take a look of MahApps.Metro.IconPacks, it's very easy to use.

For such, [SVGImage](https://github.com/dotnetprojects/SVGImage) should be enough, I could add the binding to that since I commit to that base.
